In [21]:
import pandas as pd

# Replace 'your_file.csv' with the path to your CSV file
df = pd.read_csv('datos_capa_8.csv')

# Display the first few rows of the dataframe
df

branches  releases  forks  watchers  stargazers  contributors   size  \
0           3         7      4        43          18             5   3775   
1           6        11     29        14         208            10    452   
2           6        30    134         5         595            25   6527   
3           5        30     11         3         118             4   1324   
4           2         3      0         2          16             2    731   
..        ...       ...    ...       ...         ...           ...    ...   
250         7         7      9         2          30             7  27526   
251         7         6      0         1          19             1  11410   
252        16         4     14         2          11             3   2641   
253         2        18     16         3          37             7    180   
254         1        21     11         2          32             7   2700   

     totalIssues  openIssues  totalPullRequests  ...  \
0             30           3                 30  ...   
1             30          19                 30  ...   
2             30          31                 30  ...   
3             30           1                 30  ...   
4              3           0                  2  ...   
..           ...         ...                ...  ...   
250           30           1                 30  ...   
251           30          10                 30  ...   
252           30          14                 30  ...   
253           30          11                 30  ...   
254           30           0                 30  ...   

     desc_trigram_most_common_test_http_server  \
0                                            0   
1                                            0   
2                                            0   
3                                            0   
4                                            0   
..                                         ...   
250                                          0   
251                                          0   
252                                          0   
253                                          0   
254                                          0   

     desc_trigram_most_common_tool_gener_github  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   
..                                          ...   
250                                           0   
251                                           0   
252                                           0   
253                                           0   
254                                           0   

     desc_trigram_most_common_track_code_coverag  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   
..                                           ...   
250                                            0   
251                                            0   
252                                            0   
253                                            0   
254                                            0   

     desc_trigram_most_common_turn_pull_request  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   
..                                          ...   
250                                           0   
251                                           0   
252                                           0   
253                                           0   
254    

In [ ]:
# Check the distribution of the target variable
target_column = "vp-category-equalfreq"
df[target_column].value_counts()


vp-category-equalfreq
0    135
1    120
Name: count, dtype: int64

In [23]:
# Display the first few rows for visual inspection
df.head()

branches  releases  forks  watchers  stargazers  contributors  size  \
0         3         7      4        43          18             5  3775   
1         6        11     29        14         208            10   452   
2         6        30    134         5         595            25  6527   
3         5        30     11         3         118             4  1324   
4         2         3      0         2          16             2   731   

   totalIssues  openIssues  totalPullRequests  ...  \
0           30           3                 30  ...   
1           30          19                 30  ...   
2           30          31                 30  ...   
3           30           1                 30  ...   
4            3           0                  2  ...   

   desc_trigram_most_common_test_http_server  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   desc_trigram_most_common_tool_gener_github  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   

   desc_trigram_most_common_track_code_coverag  \
0                                            0   
1                                            0   
2                                            0   
3                                            0   
4                                            0   

   desc_trigram_most_common_turn_pull_request  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   

   desc_trigram_most_common_updat_commit_push  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   

   desc_trigram_most_common_updat_yaml_properti  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   

   desc_trigram_most_common_upload_aab_file  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   desc_trigram_most_common_use_browser_api  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   desc_trigram_most_common_valid_dependabot_yaml  \
0                                               0   
1                                               0   
2                                               0   
3                                               0   
4                                               0   

   desc_trigram_most_common_yet_anoth_upload  
0                                          0  
1                                          0  
2                                          0  
3                                          0  
4                                          0  

[5 rows x 1141 columns]

In [24]:
# Check for missing values in the dataset
missing_values = df.isnull().sum()

# Display columns with missing values
missing_values[missing_values > 0]


Series([], dtype: int64)

In [25]:
# Check for duplicate rows
duplicate_rows = df.duplicated().sum()

print(f"Number of duplicate rows: {duplicate_rows}")


Number of duplicate rows: 0


In [26]:
# Display summary statistics for numerical columns
df.describe().transpose()


count        mean         std  \
branches                                        255.0    7.015686    6.828592   
releases                                        255.0   14.439216   10.363247   
forks                                           255.0   33.670588   58.449397   
watchers                                        255.0    5.764706    7.888465   
stargazers                                      255.0  184.541176  552.088395   
...                                               ...         ...         ...   
desc_trigram_most_common_updat_yaml_properti    255.0    0.003922    0.062622   
desc_trigram_most_common_upload_aab_file        255.0    0.003922    0.062622   
desc_trigram_most_common_use_browser_api        255.0    0.003922    0.062622   
desc_trigram_most_common_valid_dependabot_yaml  255.0    0.003922    0.062622   
desc_trigram_most_common_yet_anoth_upload       255.0    0.003922    0.062622   

                                                 min   25%   50%    75%  \
branches                                         1.0   2.0   5.0    8.0   
releases                                         0.0   6.0  11.0   26.5   
forks                                            0.0   6.0  13.0   29.5   
watchers                                         0.0   2.0   3.0    5.0   
stargazers                                      11.0  23.5  45.0  117.0   
...                                              ...   ...   ...    ...   
desc_trigram_most_common_updat_yaml_properti     0.0   0.0   0.0    0.0   
desc_trigram_most_common_upload_aab_file         0.0   0.0   0.0    0.0   
desc_trigram_most_common_use_browser_api         0.0   0.0   0.0    0.0   
desc_trigram_most_common_valid_dependabot_yaml   0.0   0.0   0.0    0.0   
desc_trigram_most_common_yet_anoth_upload        0.0   0.0   0.0    0.0   

                                                   max  
branches                                          30.0  
releases                                          30.0  
forks                                            367.0  
watchers                                          67.0  
stargazers                                      6351.0  
...                                                ...  
desc_trigram_most_common_updat_yaml_properti       1.0  
desc_trigram_most_common_upload_aab_file           1.0  
desc_trigram_most_common_use_browser_api           1.0  
desc_trigram_most_common_valid_dependabot_yaml     1.0  
desc_trigram_most_common_yet_anoth_upload          1.0  

[1141 rows x 8 columns]

# 📌 Observations and Next Steps

## 🔍 Key insights from the summary statistics:

- ✅ Some columns (e.g., stargazers, forks) show high variance, with stargazers having a max value of 6351 while the median is only 45. This suggests a skewed distribution with a few repositories having extreme popularity.

- ✅ Some features (e.g., desc_trigram_most_common_*) are sparse, with most values being 0.

- ✅ Certain columns (e.g., releases) contain zero values, which may or may not be meaningful (e.g., repositories with no releases).

- ✅ Features like branches, watchers, and forks have relatively low dispersion, making them potentially stable predictors.

In [27]:
# Identify outliers using the Interquartile Range (IQR) method
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR = Q3 - Q1

# Define outlier thresholds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Count how many values are considered outliers per column
outlier_counts = ((df < lower_bound) | (df > upper_bound)).sum()

# Display only columns where outliers exist
outlier_counts[outlier_counts > 0]


branches                                          24
forks                                             35
watchers                                          40
stargazers                                        38
contributors                                      18
                                                  ..
desc_trigram_most_common_updat_yaml_properti       1
desc_trigram_most_common_upload_aab_file           1
desc_trigram_most_common_use_browser_api           1
desc_trigram_most_common_valid_dependabot_yaml     1
desc_trigram_most_common_yet_anoth_upload          1
Length: 1065, dtype: int64

# Why Log Transformation?

- ✅ Reduces the impact of extreme values without losing information.

- ✅ Improves model performance by making distributions more symmetric.

- ✅ Preserves the ranking of values (i.e., a high value remains high but in a compressed range).

In [29]:
import numpy as np

# Define the selected features for log transformation
log_features = ["forks", "watchers", "stargazers", "contributors", "branches"]

# Apply log1p transformation to selected columns
df[log_features] = df[log_features].apply(lambda x: np.log1p(x))

# Verify the transformation by checking the new distribution
df[log_features].describe().transpose()


count      mean       std       min       25%       50%  \
forks         255.0  2.714755  1.261364  0.000000  1.945910  2.639057   
watchers      255.0  1.605249  0.695492  0.000000  1.098612  1.386294   
stargazers    255.0  4.131009  1.240886  2.484907  3.198465  3.828641   
contributors  255.0  2.003699  0.677397  0.693147  1.386294  1.945910   
branches      255.0  1.781131  0.766266  0.693147  1.098612  1.791759   

                   75%       max  
forks         3.417592  5.908083  
watchers      1.791759  4.219508  
stargazers    4.770649  8.756525  
contributors  2.484907  3.433987  
branches      2.197225  3.433987

In [30]:
df.to_csv('data.csv', index=False)